packages/relay/src/Environment.js

```diff
@@ -1,13 +1,54 @@
 // @flow
 
 import { GetDeviceLocale } from '@kiwicom/mobile-localization';
-import { Environment, Network, RecordSource, Store } from 'relay-runtime';
+import {
+  Environment,
+  Network,
+  RecordSource,
+  Store,
+  Observable,
+  createOperationSelector,
+} from 'relay-runtime';
+import { AsyncStorage } from 'react-native';
 
 type FetcherResponse = {|
   +data: Object,
   +errors?: $ReadOnlyArray<Object>,
 |};
 
+const ASYNC_STORE_KEY = '@OfflineStore:key';
+const GRAPHQL_URL = 'https://graphql.kiwi.com/';
+
+const store = new Store(new RecordSource());
+
+function fetchFromTheNetwork(networkHeaders, operation, variables, observer) {
+  fetch(GRAPHQL_URL, {
+    method: 'POST',
+    headers: networkHeaders,
+    body: JSON.stringify({
+      query: operation.text, // TODO: fetch persisted queries instead (based on operation.id)
+      variables,
+    }),
+  })
+    .then(fetchResponse => fetchResponse.json())
+    .then(jsonResponse => {
+      if (jsonResponse.errors) {
+        jsonResponse.errors.forEach(error => console.warn(error));
+      }
+      observer.next(jsonResponse);
+      observer.complete();
+    })
+    .then(() => {
+      AsyncStorage.setItem(
+        ASYNC_STORE_KEY,
+        JSON.stringify(store.getSource()),
```

Comment:
And here, you're setting the key `ASYNC_STORE_KEY` with a whole dump of the store.

Reply:
Yes, the whole `store` snapshot will be stored to the `AsyncStorage` so I can read it with the next request and use it immediately without waiting for the network. That's the biggest improvement and UI should feel more responsive now (not long loading spinners)...